packages/expo-app-loading/src/AppLoading.tsx

```diff
@@ -0,0 +1,69 @@
+import React from 'react';
+
+import { _emitEvent } from './LifecycleEmitter';
+import * as SplashScreen from './SplashScreen';
+
+export { getAppLoadingLifecycleEmitter } from './LifecycleEmitter';
+
+const AppLoading: React.FC<AppLoadingProps> = props => {
+  const startLoadingResources = React.useCallback(async () => {
+    if (!props.startAsync) {
+      return;
+    }
+
+    if (!props.onFinish) {
+      throw new Error('AppLoading onFinish prop is required if startAsync is provided');
+    }
+
+    try {
+      await props.startAsync();
+    } catch (error) {
+      props.onError!(error);
```

Comment:
It's prefilled by default with `defaultProps`. I can move the default error handling to this one too if you need it. But because this is prefilled, it's optional on the public API of the component. Typescript doesn't care if it's prefilled or not, so this is one of the few scenarios where I think it's ok-ish to use this. The only downside is when the user manually set `onError` to undefined, e.g. `<AppLoading onError={undefined} />`.

Reply:
Do you think we should throw an error if there is no valid `onError` handler? Something like `props.onError.?(error)` might be more correct then.